src/main/java/ru/ralsei/whatcanyousee/logic/MazeGameMap.java

```diff
@@ -0,0 +1,809 @@
+         */
+        private int distance = 0;
+
+        /**
+         * If lower than map.distanceId, distance of this cell is invalid and should be updated.
+         */
+        private int distanceId = 0;
+
+        /**
+         * Current cell's image.
+         */
+        private int image = R.drawable.emptycell;
+
+        /**
+         * Default cell's image.
+         */
+        private int defaultImage = R.drawable.emptycell;
+
+        public void setToogle(Toogle toogle) {
+            this.toogle = toogle;
+        }
+
+        public void setTrap(Trap trap) {
+            this.trap = trap;
+        }
+
+        public void setImage(int image) {
+            this.image = image;
+        }
+
+        public void setDefaultImage(int image) {
+            this.defaultImage = image;
+        }
+
+        void resetImage() {
+            this.image = defaultImage;
+        }
+
+        boolean isWall() {
+            return isWall;
+        }
+
+        boolean isTrap() {
+            return trap != null;
+        }
+
+        boolean isToogle() {
+            return toogle != null;
+        }
+
+        Trap getTrap() {
+            return trap;
+        }
+
+        Toogle getToogle() {
+            return toogle;
+        }
+
+        /**
+         * Makes given cell a wall cell.
+         */
+        public void makeWall() {
+            isWall = true;
+            defaultImage = R.drawable.wall;
+            image = R.drawable.wall;
+        }
+
+        /**
+         * Make given cell not a wall.
+         */
+        public void makeNotWall() {
+            isWall = false;
+            defaultImage = R.drawable.emptycell;
+            image = R.drawable.emptycell;
+        }
+
+        public int getDistance() {
+            return distance;
+        }
+
+        void setDistance(int distance) {
+            this.distance = distance;
+        }
+
+        int getX() {
+            return x;
+        }
+
+        private void setX(int x) {
+            this.x = x;
+        }
+
+        int getY() {
+            return y;
+        }
+
+        private void setY(int y) {
+            this.y = y;
+        }
+
+        int getDistanceId() {
+            return distanceId;
+        }
+
+        void setDistanceId(int distanceId) {
+            this.distanceId = distanceId;
+        }
+    }
+
+    /**
+     * Returns current player's coordinates.
+     */
+    Coordinates getCurrentCoordinates() {
+        return currentCoordinates;
+    }
+
+    /**
+     * Sets current player's coordinates to another.
+     */
+    void setCurrentCoordinates(Coordinates coordinates) {
+        this.currentCoordinates = coordinates;
+    }
+
+    /*
+    private final Object gameResultLock = new Object();
+
+    Object getGameResultLock() {
+        return gameResultLock;
+    }
+    */
+
+    /**
+     * If game is over, shows if player either won or lost.
+     */
+    private volatile Boolean isWin = null;
```

Comment:
I suggest using Optional<Boolean> instead of a Boolean that can take a null value. This is more safe with respect to accidentally introducing a NullPointerException.

Reply:
Optional does not seems to work well as volatile (or maybe I did something wrong), and also gives warning "Optional used as type for field".
The point was to replace synchronized blocks with a volatile variable for better performance, but if it's not OK, I better just use two booleans.

Reply:
Regarding Optional, you are right. But the code still stinks a bit. How about an Enum to denote {game in progress, lost, won} state?

Reply:
Done.